lib/Frontend/Frontend.cpp

```diff
@@ -804,13 +804,14 @@ SourceFile *CompilerInstance::createSourceFileForMainModule(
     SourceFileKind fileKind, SourceFile::ImplicitModuleImportKind importKind,
     Optional<unsigned> bufferID) {
   ModuleDecl *mainModule = getMainModule();
+  bool isPrimary = bufferID && isPrimaryInput(*bufferID);
   SourceFile *inputFile = new (*Context)
       SourceFile(*mainModule, fileKind, bufferID, importKind,
                  Invocation.getLangOptions().CollectParsedToken,
-                 Invocation.getLangOptions().BuildSyntaxTree);
+                 Invocation.getLangOptions().BuildSyntaxTree, isPrimary);
   MainModule->addFile(*inputFile);
 
-  if (bufferID && isPrimaryInput(*bufferID)) {
+  if (isPrimary) {
```

Comment:
Why is this change needed?

Reply:
As far as I can tell,  this is creating redundant information about primaryness. Is there some reason to redundantly store the boolean in the SourceFile?

Reply:
I think the crux of the issue is that we're creating a SILProfiler per SILFunction, and SILFunctions are used for declarations as well as definitions. For a declaration, we should not be creating a SILProfiler, nor should we be walking the AST of the function.

Reply:
Thanks, I'll look into this!